cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -179,20 +182,29 @@ func buildSecureBootstrapKubeConfig(endpoint string, caCert []byte) *clientcmdap
 }
 
 // runForEndpointsAndReturnFirst loops the endpoints slice and let's the endpoints race for connecting to the master
-func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(string) (*clientcmdapi.Config, error)) *clientcmdapi.Config {
+func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(string) (*clientcmdapi.Config, error)) (*clientcmdapi.Config, error) {
 	stopChan := make(chan struct{})
+	const retries = int(constants.DiscoveryTimeout / constants.DiscoveryRetryInterval)
 	var resultingKubeConfig *clientcmdapi.Config
 	var once sync.Once
 	var wg sync.WaitGroup
+	var resultingError error
 	for _, endpoint := range endpoints {
 		wg.Add(1)
 		go func(apiEndpoint string) {
+			retryCount := 1
 			defer wg.Done()
 			wait.Until(func() {
-				fmt.Printf("[discovery] Trying to connect to API Server %q\n", apiEndpoint)
+				fmt.Printf("[discovery] Trying to connect to API Server %q (%d of %d tries)\n", apiEndpoint, retryCount, retries)
 				cfg, err := fetchKubeConfigFunc(apiEndpoint)
 				if err != nil {
 					fmt.Printf("[discovery] Failed to connect to API Server %q: %v\n", apiEndpoint, err)
+					retryCount++
+					if retryCount > retries {
+						resultingError = err
+						close(stopChan)
+						fmt.Printf("[discovery] Abort connecting to API Server %q after %d retries: %v\n", apiEndpoint, retries, err)
```

Comment:
How about `resultingError = fmt.Errorf("abort connecting to API Server %q after %d retries: %v\n", apiEndpoint, retries, err)`?